File: cola2_lib_ros/python/cola2_ros/transform_handler.py
#!/usr/bin/env python
# Copyright (c) 2020 Iqua Robotics SL - All Rights Reserved
#
# This file is subject to the terms and conditions defined in file
# 'LICENSE.txt', which is part of this source code package.

"""
Module to get transform between different frames.
"""

from __future__ import print_function
import rospy
import tf
from numpy import rad2deg


class TransformHandler(object):
    """
    The TransformHandler class queries and saves transforms with origin at the vehicle frame.
    """
    def __init__(self):
        """
        Constructor that queries the namespace to know vehicle frame."""
        self.transforms = dict()  # transforms from the robot to the sensors
        self.tf_listener = tf.TransformListener()  # transform listener
        self.frame_vehicle = rospy.get_namespace() + "base_link"  # vehicle frame

    def get_transform(self, frame):
        """
        Get a static transform from the map or query the listener and save.

        :param frame: Frame name.
        :type frame: str
        :return: Returns a tuple containig if call was successfull and list with the transform in xyzrpy.
        :rtype: tuple[bool,list[float,float,float],list[float,float,float]]
        """
        # Return identity when frame is vehicle frame
        if frame == self.frame_vehicle:
            return True, [0., 0., 0.], [0., 0., 0.]
        # Look for the transform
        if frame in self.transforms:
            # Found in map
            xyz, rpy = self.transforms[frame]
            return True, xyz, rpy
        else:
            # Need to query it
            ok, xyz, rpy = self.get_dynamic_transform(frame)
            if ok:
                self.transforms[frame] = (xyz, rpy)
                rospy.loginfo("Transform Handler added: {:s}".format(frame))
                rospy.loginfo("trans: {:.3f} {:.3f} {:.3f}".format(*xyz))
                rpy_deg = [rad2deg(v) for v in rpy]
                rospy.loginfo("rpy_deg: {:.3f} {:.3f} {:.3f}".format(*rpy_deg))
                return True, xyz, rpy
        return False, [0., 0., 0.], [0., 0., 0.]

    def get_dynamic_transform(self, frame):
        """
        Get a transform by querying the listener.

        :param frame: Frame name.
        :type frame: str
        :return: Returns a tuple containig if call was successfull and list with the transform in xyzrpy.
        :rtype: tuple[bool,list[float,float,float],list[float,float,float]]
        """
        # Return identity when frame is vehicle frame
        if frame == self.frame_vehicle:
            return True, [0., 0., 0.], [0., 0., 0.]
        # Look for the transform
        try:
            self.tf_listener.waitForTransform(self.frame_vehicle, frame, rospy.Time(0),rospy.Duration(0.5))
            (xyz, quat) = self.tf_listener.lookupTransform(self.frame_vehicle, frame, rospy.Time(0))
            rpy = tf.transformations.euler_from_quaternion(quat)
            return True, xyz, rpy
        except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
            rospy.logfatal("Unable to get dynamic transform: {:s}".format(frame))
            return False, [0., 0., 0.], [0., 0., 0.]

if __name__ == '__main__':
    # init
    rospy.init_node('transform_handler_test')
    node = TransformHandler()
    print("frame vehicle: {:s}".format(node.frame_vehicle))

    # test
    node.get_transform("base_link")
    node.get_transform("adis_imu")
    node.get_dynamic_transform("adis_imu")

File: cola2_lib_ros/src/cola2_lib_ros/transform_handler.cpp
/*
 * Copyright (c) 2020 Iqua Robotics SL - All Rights Reserved
 *
 * This file is subject to the terms and conditions defined in file
 * 'LICENSE.txt', which is part of this source code package.
 */

#include "cola2_lib_ros/transform_handler.h"

cola2::ros::TransformHandler::TransformHandler() : tf_buffer_(), tf_listener_(tf_buffer_)
{
  base_frame_ = cola2::ros::getNamespaceNoInitialDash() + std::string("/base_link");
}

cola2::ros::TransformHandler::TransformHandler(const std::string& frame) : tf_buffer_(), tf_listener_(tf_buffer_)
{
  base_frame_ = frame;
}

bool cola2::ros::TransformHandler::getTransform(const std::string& frame, Eigen::Isometry3d& transform)
{
  // Cleanup / at begining of frame (workaround, should not happen)
  std::string cleaned_frame = frame;
  if (frame[0] == '/')
  {
    cleaned_frame = frame.substr(1, frame.size() - 1);
  }

  // Return identity when frame is vehicle frame
  if (cleaned_frame.compare(base_frame_) == 0)
  {
    transform = Eigen::Isometry3d::Identity();
    return true;
  }
  // Look for the transform
  try
  {
    // Found in map
    transform = transforms_.at(cleaned_frame);
    return true;
  }
  catch (const std::out_of_range&)
  {
    // Need to query it
    if (getDynamicTransform(cleaned_frame, transform))
    {
      transforms_[cleaned_frame] = transform;
      ROS_INFO_STREAM("Transform Handler added '" << cleaned_frame);
      ROS_INFO_STREAM("trans: " << transform.translation().transpose());
      ROS_INFO_STREAM("rpy:   " << cola2::utils::rotation2euler(transform.rotation()).transpose());
      return true;
    }
  }
  return false;
}

bool cola2::ros::TransformHandler::getDynamicTransform(const std::string& frame, Eigen::Isometry3d& transform,
                                                       const ::ros::Time& time) const
{
  // Return identity when frame is vehicle frame
  if (frame.compare(base_frame_) == 0)
  {
    transform = Eigen::Isometry3d::Identity();
    return true;
  }
  // Look for the transform
  try
  {
    const geometry_msgs::TransformStamped tf_trans = tf_buffer_.lookupTransform(base_frame_, frame, time);
    transform = tfTransformToEigen(tf_trans);
    return true;
  }
  catch (tf2::TransformException& ex)
  {
    ROS_FATAL_STREAM("Unable to get dynamic transform: " << frame << "\n  - " << std::string(ex.what()));
  }
  return false;
}

Eigen::Isometry3d cola2::ros::TransformHandler::tfTransformToEigen(const geometry_msgs::TransformStamped& trans) const
{
  const Eigen::Quaterniond quat(trans.transform.rotation.w, trans.transform.rotation.x, trans.transform.rotation.y,
                                trans.transform.rotation.z);
  Eigen::Isometry3d affine(quat.toRotationMatrix());
  affine.translation() =
      Eigen::Vector3d(trans.transform.translation.x, trans.transform.translation.y, trans.transform.translation.z);
  return affine;
}

void cola2::ros::TransformHandler::setTransformManually(const std::string& frame, const double x, const double y,
                                                        const double z, const double qx, const double qy,
                                                        const double qz, const double qw)
{
  Eigen::Quaterniond quat(qw, qx, qy, qz);
  Eigen::Isometry3d affine(quat.toRotationMatrix());
  affine.translation() = Eigen::Vector3d(x, y, z);
  transforms_[frame] = affine;
}

// ===========================
// DEPRECATED
// ===========================
bool cola2::ros::TransformHandler::getTransform(const std::string& frame, Eigen::Affine3d& transform)
{
  // call with isometry
  Eigen::Isometry3d trans;
  const bool success = getTransform(frame, trans);
  // transform to affine
  // transform = trans.affine();
  transform = Eigen::Affine3d(trans.rotation());
  transform.translation() = trans.translation();
  return success;
}

bool cola2::ros::TransformHandler::getDynamicTransform(const std::string& frame, Eigen::Affine3d& transform,
                                                       const ::ros::Time& time) const
{
  // call with isometry
  Eigen::Isometry3d trans;
  const bool success = getDynamicTransform(frame, trans, time);
  // transform to affine
  // transform = trans.affine();
  transform = Eigen::Affine3d(trans.rotation());
  transform.translation() = trans.translation();
  return success;
}
